mne/beamformer/_lcmv.py

```diff
@@ -101,54 +135,69 @@ def _apply_lcmv(data, info, tmin, forward, noise_cov, data_cov, reg,
-                if i != eig_vals.argmax() and i != eig_vals.argmin():
-                    idx_middle = i
-
-            # TODO: The eigenvector associated with the smallest eigenvalue
-            # should probably be used when using combined EEG and MEG data
-            max_ori = eig_vecs[:, idx_middle]
-
-            Wk[:] = np.dot(max_ori, Wk)
-            Ck = np.dot(max_ori, np.dot(Ck, max_ori))
-            is_free_ori = False
-
-        if is_free_ori:
-            # Free source orientation
-            Wk[:] = np.dot(linalg.pinv(Ck, 0.1), Wk)
-        else:
-            # Fixed source orientation
-            Wk /= Ck
+            # weight normalization and orientation selection:
+            if weight_norm is not None and pick_ori == 'max-power':
+                # finding optimal orientation for NAI and unit-noise-gain
+                # based on [2]_, Eq. 4.47
+                tmp = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+                eig_vals, eig_vecs = linalg.eig(np.dot(linalg.pinv(tmp),
+                                                       np.dot(Wk, Gk)))
+                idx_max = eig_vals.argmax()
+                max_ori = eig_vecs[:, idx_max]
+                Wk[:] = np.dot(max_ori, Wk)
+                Gk = np.dot(Gk, max_ori)
+
+                # compute spatial filter for NAI or unit-noise-gain
+                tmp = np.dot(Gk.T, np.dot(Cm_inv_sq, Gk))
+                denom = np.sqrt(tmp)
+                Wk /= denom
+                if weight_norm == 'nai':
+                    Wk /= np.sqrt(noise)
+
+                is_free_ori = False
+
+            # no weight-normalization and max-power is not implemented yet:
+            else:
+                raise ValueError('The max-power orientation selection is not '
+                                 'yet implemented with weight_norm set to'
+                                 'None.')
+
+        else:  # do vector beamformer
+            # compute the filters:
+            if is_free_ori:
+                # Free source orientation
+                Wk[:] = np.dot(linalg.pinv(Ck, 0.1), Wk)
+            else:
+                # Fixed source orientation
+                Wk /= Ck
+
+            # handle noise normalization with free/normal source orientation:
+            if weight_norm == 'nai':
+                raise ValueError('Weight normalization with neural activity '
+                                 'index is not implemented yet with free or '
+                                 'fixed orientation.')
+
+            if weight_norm == 'unit-noise-gain':
+                noise_norm = np.sum(Wk ** 2, axis=1)
+                if is_free_ori:
+                    noise_norm = np.sum(noise_norm)
+                noise_norm = np.sqrt(noise_norm)
+                if noise_norm == 0.:
+                    noise_norm_inv = 0  # avoid division by 0
+                else:
+                    noise_norm_inv = 1. / noise_norm
+                Wk[:] *= noise_norm_inv
```

Comment:
Mh, this is the original weight normalization that is now applied during the weight calculation instead of at the end - I think the same thing happens in the old version as well. And if I see it right it only should happen if Wk was 0 anyway (compare line 184-187), no?